src/spider/client/Data.hpp

```diff
@@ -0,0 +1,81 @@
+#ifndef SPIDER_CLIENT_DATA_HPP
+#define SPIDER_CLIENT_DATA_HPP
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace spider {
+
+class DataImpl;
+
+template <class T>
+class Data {
+private:
+    std::unique_ptr<DataImpl> m_impl;
+
+public:
+    /**
+     * Gets the values stored in Data.
+     * @return value stored in Data.
+     */
+    auto get() -> T;
+    /**
+     * Indicates that the data is persisted and should not be rollbacked
+     * on failure recovery.
+     */
+    // Not implemented in milestone 1
+    // void mark_persist();
+    /**
+     * Sets locality list of the data.
+     * @param nodes nodes that has locality
+     * @param hard true if the locality list is a hard requirement, false otherwise
+     */
+    void set_locality(std::vector<std::string> const& nodes, bool hard);
+
+    class Builder {
+    private:
+    public:
+        /**
+         * Sets the key for the data. If no key is provided, Spider generates a key.
```

Comment:
If the key is optional, how about using `std::optional` to make that explicit in the API itself?

Reply:
Sorry. Wrong docstring. Spider doesn't generate a key if no key is provided.
The point of using builder pattern is to provide flexibility so that users don't call the function they don't need.